.github/scripts/replace-mmtk-dep.py

```diff
@@ -2,35 +2,74 @@
 import tomlkit
 
 parser = argparse.ArgumentParser(
         description='Replace the mmtk-core dependency of a given VM binding',
         )
 
 parser.add_argument('toml_path', help='Path to Cargo.toml')
-parser.add_argument('mmtk_core_path', help='Path to the mmtk_core repo')
+# The following arguments are exclusive. Use either. If both are supplied, we use the local path.
+# 1. Point to a local path
+parser.add_argument('--mmtk_core_path', help='Path to the mmtk_core repo.')
+# 2. Point to a remote repo
+parser.add_argument('--mmtk_core_git', help='URL to the mmtk_core repo.')
+parser.add_argument('--mmtk_core_rev', help='Revision to use')
 
 args = parser.parse_args()
 
+# Check what we should do.
+if args.mmtk_core_path is not None:
+    how = "point_to_local"
+elif args.mmtk_core_git is not None and args.mmtk_core_rev is not None:
+    how = "point_to_repo"
+else:
+    print("No path or git/rev is supplied. We cannot update the toml")
+    sys.exit(1)
+
 print("Reading TOML from '{}'".format(args.toml_path))
 with open(args.toml_path, "rt") as f:
     toml_data = tomlkit.load(f)
 
+if "mmtk" not in toml_data["dependencies"]:
+    print("Cannot find the mmtk dependency in {}".format(args.toml_path))
+    sys.exit(1)
+
+# The mmtk dependency could be an inlined table for some bindings:
+# [dependencies]
+# mmtk = { git = "...", rev = "..." }
+# But it could be a subtable for other bindings:
+# [dependencies.mmtk]
+# git = "..."
+# rev = "..."
 mmtk_node = toml_data["dependencies"]["mmtk"]
 
-# These keys may specify the locations of the dependency. Remove them.
-for key in ["git", "branch", "version", "registry"]:
-    if key in mmtk_node:
-        print("Deleting dependencies.mmtk.{}".format(key))
-        del mmtk_node[key]
-    else:
-        print("Key dependencies.mmtk.{} does not exist.  Ignored.".format(key))
-
-# Use mmtk-core from the specified local directory.
-mmtk_repo_path = os.path.realpath(args.mmtk_core_path)
-print("Setting dependencies.mmtk.path to {}".format(mmtk_repo_path))
-mmtk_node["path"] = mmtk_repo_path
+def remove_keys(item, keys):
+    for key in keys:
+        if key in item:
+            print("Deleting dependencies.mmtk.{}".format(key))
+            del item[key]
+        else:
+            print("Key dependencies.mmtk.{} does not exist.  Ignored.".format(key))
+
+if how == "point_to_local":
+    # Remove any existing key about the versioning. We use whatever version we have in the local path.
+    remove_keys(mmtk_node, ["git", "branch", "registry", "rev"])
+
+    # Use mmtk-core from the specified local directory.
+    mmtk_repo_path = os.path.realpath(args.mmtk_core_path)
+    print("Setting dependencies.mmtk.path to {}".format(mmtk_repo_path))
+    mmtk_node["path"] = mmtk_repo_path
+elif how == "point_to_repo":
+    # Remove any existing key about local path
+    remove_keys(mmtk_node, ["path"])
+    # Update git/rev
+    mmtk_node["git"] = args.mmtk_core_git
+    mmtk_node["rev"] = args.mmtk_core_rev
```

Comment:
I think we should remove all of the keys we know, regardless weather it is "point_to_local" or "point_to_repo", just in case any of them is defined and we are not aware of it.  For example, in the case of "point_to_repo", it does not make sense if more than one of "branch", "tag" and "rev" are specified.

```suggestion
remove_keys(mmtk_node, ["git", "rev", "tag", "branch", "path", "registry", "version"])
if how == "point_to_local":
    # Use mmtk-core from the specified local directory.
    mmtk_repo_path = os.path.realpath(args.mmtk_core_path)
    print("Setting dependencies.mmtk.path to {}".format(mmtk_repo_path))
    mmtk_node["path"] = mmtk_repo_path
elif how == "point_to_repo":
    # Update git/rev
    mmtk_node["git"] = args.mmtk_core_git
    mmtk_node["rev"] = args.mmtk_core_rev
```

Reply:
I just realised that this script is used for the purpose of making permanent changes to the binding's `Cargo.toml` instead of temporarily changing the dependency to test against a specific version of mmtk-core.

In this case, formatting matters.  There may be no general way to change `Cargo.toml` "back", because the style of the original `Cargo.toml` before setting `git` and `rev` to a specific repo and a specific rev may still vary from binding to binding.

mmtk-openjdk has this:

```toml
# But other changes including adding/removing whitespaces in commented lines may break the CI.
mmtk = { git = "https://github.com/mmtk/mmtk-core.git", rev = "9a676e68dd41272b57e6812dc404ffa6391668e2" }
# Uncomment the following to build locally
# mmtk = { path = "../repos/mmtk-core" }
```

but mmtk-ruby has this:

```toml
[dependencies.mmtk]
features = ["is_mmtk_object", "object_pinning"]

# Uncomment one of the following lines to choose where to find mmtk-core.
git = "https://github.com/mmtk/mmtk-core.git"  # Use mmtk-core from the official repository.
#path = "../../mmtk-core"  # Use mmtk-core from a local repository.
```

I think it is OK to assume our officially supported bindings have `git` and `rev` set so that we only need to update those.  `mmtk-ruby` currently doesn't have the `rev` field, but when running `cargo build`, it still fetches the revision specified in `Cargo.lock`.  I would argue that having the revision in `Cargo.lock` is sufficient for tracking the version we are testing against, but it is harmless to let `mmtk-ruby` have the `rev` field in the `mmtk-core` dependency, too.

Reply:
If I understand you right:
* If we are setting a path, we remove all the known keys about versioning, and add `path`. This will mess up the formatting. But the change is not persistent, so it doesn't matter.
* If we are updating `git` and `rev`, the change will be persistent and we want to preserve formatting. So we do not remove keys, and just update those two fields. This preserves the formatting.

I made changes as above. Let me know if I misunderstood any part.

Reply:
Yes.  That's what I meant.